torch/fx/experimental/symbolic_shapes.py

```diff
@@ -186,79 +176,217 @@ def magic_impl(self, other):
+    @_lru_cache
+    def replace(self, expr: "sympy.Expr") -> "sympy.Expr":
+        replacements = {s: self._find(s) for s in expr.free_symbols}
+        return sympy.expand(expr.xreplace(replacements))
+
+    @_lru_cache
+    def _update_divisible(self):
+        new_divisible = {}
+        for k in self.divisible:
+            res = self.replace(k)
+            if len(res.free_symbols) > 0:
+                new_divisible[k] = 0
+
+        self.divisible = new_divisible
+
+    @_lru_cache
+    def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
+        expr = self.replace(expr)
+        if len(expr.atoms(sympy.Mod)) > 0:
+            self._update_divisible()
+            expr = expr.xreplace(self.divisible)
+            expr = sympy.expand(expr)
+        return expr
+
+    @lru_cache(256)
+    def size_hint(self, expr: "sympy.Expr"):
+        result_expr = sympy.expand(expr).xreplace(self.var_to_val)
+        assert len(result_expr.free_symbols) == 0, "Size hint has variables we don't have underlying values for"
+        return result_expr
+
+    def _find(self, a):
+        """
+        Implements a DSU to find the variable that represents a
+        TODO: Improve this to handle non-identity transitive replacements
+        """
+        acopy = a
+        while a in self.replacements:
+            a = self.replacements[a]
+        while acopy in self.replacements:
+            self.replacements[acopy], acopy = a, self.replacements[acopy]
+        return a
+
+    def _maybe_guard_eq(self, expr: "sympy.Eq") -> None:
+        """
+        Evaluates the result of an eq call. If true, uses information to
+        simplify shapes (i.e. a == b or a % 5 == 0)
+        """
+        concrete_bool = bool(self.size_hint(expr))
+        if not concrete_bool:
+            return
+        free = list(expr.free_symbols)
+
+        assert len(free) > 0, "The expression should not be static by this point"
+        # In case of really gnarly expression, we don't blow up
+        if len(free) <= 4:
+            free = sorted(free, key=lambda x: (self.size_hint(x), x.name), reverse=True)  # type: ignore[attr-defined]
+            lhs = expr.lhs
+            rhs = expr.rhs
+            try:
+                solutions = sympy.solveset(lhs - rhs, free[0], domain=sympy.S.Integers)
+                if not solutions.is_finite_set:
+                    if len(expr.atoms(sympy.Mod)) == 1:
+                        mod_expr = tuple(expr.atoms(sympy.Mod))[0]
+                        solutions = sympy.solveset(lhs - rhs, mod_expr, domain=sympy.S.Integers)
+                        if solutions.is_finite_set and len(solutions) == 1 and tuple(solutions)[0] == 0:
+                            self.divisible[mod_expr] = 0
+                    return
+
+                if not isinstance(solutions, sympy.FiniteSet):
+                    return
+
+                solutions = tuple(solutions)
+                if len(solutions) == 1 and "/" not in str(solutions[0]):
```

Comment:
Stringification police! Ask the sympy devs how to do this better lol

Reply:
Actually I know how to do this better now lol. You can just check atoms.